### PR TITLE
[jit] Let script module buffer attributes can also cast device/type

### LINKED ISSUE #19039: Cannot change device of JIT module after initial loading using C++ frontend
## 🐛 Bug

I exported a JIT model on GPU, then try to load it and run it on CPU, using C++ frontend. 

## To Reproduce

1. Create a demo model and export it on GPU
```
import torch

class BatchNormModel(torch.jit.ScriptModule):
    def __init__(self):
        super(BatchNormModel, self).__init__(optimize=False)
        self.bn = torch.nn.BatchNorm1d(10)

    @torch.jit.script_method
    def forward(self, x):
        return self.bn(x)


model = BatchNormModel()
model.eval()
model.cuda()
model.save("/tmp/bn_gpu.pt")
```
2. Try to load the model and execute on CPU
```
#include <torch/script.h>

#include <memory>

int main() {
    std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("/tmp/bn_gpu.pt");
    module->to(torch::kCPU);

    std::vector<torch::jit::IValue> inputs;
    inputs.emplace_back(torch::randn({6, 10}));
    auto output = module->forward(inputs).toTensor();
}
```
Error
```
Terminate std::runtime_error( Expected tensor to have CPU Backend, but got tensor with CUDA Backend (while checking arguments for batch_norm_cpu) (checkBackend at ../aten/src/ATen/TensorUtils.cpp:202)
```

## Expected behavior

Expect `->to` function to recursively move all the parameters on CPU.

## Environment

Please copy and paste the output from our
PyTorch version: 1.1.0a0+9696f06
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 14.04.6 LTS
GCC version: (Ubuntu 4.8.4-2ubuntu1~14.04.4) 4.8.4
CMake version: version 3.11.1

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: TITAN V
GPU 1: TITAN Xp
GPU 2: GeForce RTX 2080 Ti

Nvidia driver version: 410.78
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.5.1.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21

Versions of relevant libraries:
[pip] numpy==1.13.3
[pip] pyarrow==0.10.0+numpy1.13.1
[conda] Could not collect

## Additional context

I have seen similar behavior when trying to cast weights to be half (FP16). If I pass a `torch::Device` object to `torch::jit::load`, model is correctly loaded on CPU so the issue seems to be related to the `->to` function.

### PR DESCRIPTION
**Pull request by wanchaol**
this fix #19039

### CODE REVIEW

**Comment by eellison** [comment]
Take a look at test/cpp/api/jit.cpp, pretty sure you can create a scriptmodule in cpp

**Comment by wanchaol** [comment]
> Take a look at test/cpp/api/jit.cpp, pretty sure you can create a scriptmodule in cpp

@eellison I was looking at this file, it can let you create a ScripModule from the function text using `torch::jit::compile`, but it could not let you defined a ScriptModule class manually. In the case of the issue, we could only trigger this case when we have a ScriptModule class in C++, where it could not be created directly in C++ (only via python + torch.jit.save + load)

**Comment by eellison** [comment]
In cpp you can define a script module with text, save it, and load it. Not sure i'm understanding how this doesn't cover your use case.

File: test/cpp/api/jit.cpp [hunk 1]
```diff
@@ -99,6 +100,19 @@ TEST(TorchScriptTest, TestDictArgMatching) {
   ASSERT_EQ(1, output.toTensor()[0].item<int64_t>());
 }
 
+
+TEST(TorchScriptTest, Conversion_MultiCUDA) {
```
Comment (driazati) [inline test/cpp/api/jit.cpp]: The test should be in `cpp/jit/`, not `api/jit.cpp`, see `testModuleDefine` to see an example of a manually added parameter (attributes have pretty much the same API on modules)

File: test/cpp/api/jit.cpp [hunk 2]
```diff
@@ -99,6 +100,19 @@ TEST(TorchScriptTest, TestDictArgMatching) {
...
+    auto module = torch::jit::load("bn_gpu.pt");
```
Comment (eellison) [inline test/cpp/api/jit.cpp]: Could you please add a run of it with GPU tensors first ?

**Review by eellison — DISMISSED** [review]
Looks good! thanks for adding the test.

File: torch/csrc/jit/script/module.cpp [hunk 3]
```diff
@@ -120,6 +120,20 @@ void Module::to_impl(
         non_blocking);
     variable.set_data(new_data);
   }
+  // Then convert every tensor attributes (buffers).
+  for (auto& parameter : get_attributes()) {
+    if (parameter.type()->isSubtypeOf(TensorType::get())) {
+      // Need to access the `at::Tensor` as a `Variable` here.
+      autograd::Variable variable = parameter.value().toTensor();
```
Comment (driazati) [inline torch/csrc/jit/script/module.cpp]: This should be moved out to a function so it's not just a copy-paste of the above code for parameters

**Review by driazati — CHANGES_REQUESTED** [review]
I don't think we should be committing binaries to the repo to be used in tests

**Review by driazati — APPROVED** [review]

**Pull request closed without merging**